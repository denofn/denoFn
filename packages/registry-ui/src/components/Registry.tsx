import { css } from "otion";
import React from "react";
import CreatableSelect from "react-select/creatable";

import { Action } from "./Function";
import { RegistryJSON } from "../types";

type Option = {
  label: string;
  value: string;
};

function createOption(label: string): Option {
  return { label, value: label };
}

function extractValues(options: Option[]): string[] {
  return (options ?? []).map(({ value }) => value);
}

export const Registry = React.memo(function Registry({
  dispatch,
  registry,
}: {
  registry?: RegistryJSON;
  dispatch: (a: Action) => void;
}) {
  const [inputValue, setInputValue] = React.useState("");
  const [options, setOptions] = React.useState(registry?.whitelist.map(createOption));

  function onClickCheckbox(e: React.ChangeEvent<HTMLInputElement>) {
    dispatch({ type: "warmupOnStart", value: !registry?.warmupOnStart });
  }

  function handleChange(options: any) {
    setOptions(options);
    dispatch({ type: "whitelist", value: extractValues(options) });
  }

  function handleKeyDown(event: React.KeyboardEvent<HTMLElement>) {
    if (!inputValue) return;
    switch (event.key) {
      case "Enter":
      case "Tab":
        handleChange([...(options ?? []), createOption(inputValue)]);
        setInputValue("");
        event.preventDefault();
    }
  }

  function handleInputChange(v: string) {
    setInputValue(v);
  }

  return (
    <div
      className={css({
        margin: "1.5rem 0",
      })}
    >
      <div className={css({ fontWeight: "bold", marginBottom: "0.75rem" })}>URL Whitelist</div>
      <CreatableSelect
        components={{ DropdownIndicator: null }}
        inputValue={inputValue}
        isClearable
        isMulti
        menuIsOpen={false}
        placeholder="Enter the domains you want to send requests to here..."
        value={options}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
        onInputChange={handleInputChange}
      />
      <div className={css({ marginTop: "1.5rem", display: "flex", alignItems: "center" })}>
        <label htmlFor="warmupOnStart" className={css({ marginRight: ".5rem", fontWeight: "bold" })}>
          Post-registration warmup
        </label>
        <input
          type="checkbox"
          checked={registry?.warmupOnStart ?? false}
          id="warmupOnStart"
          onChange={onClickCheckbox}
        />
      </div>
    </div>
  );
});
