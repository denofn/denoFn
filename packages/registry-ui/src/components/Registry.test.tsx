import { render, cleanup } from "@testing-library/react";
import React from "react";

import { Registry } from "./Registry";

afterEach(cleanup);

it("renders with the correct styles", () => {
  const { asFragment } = render(<Registry dispatch={() => {}} />);
  expect(asFragment()).toMatchSnapshot();
});
