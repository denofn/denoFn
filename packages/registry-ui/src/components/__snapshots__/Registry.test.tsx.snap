// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with the correct styles 1`] = `
<DocumentFragment>
  <div
    class="_di48ht"
  >
    <div
      class="_1efi8gv _1hl83va"
    >
      URL Whitelist
    </div>
    <div
      class=" css-2b097c-container"
    >
      <div
        class=" css-yk16xz-control"
      >
        <div
          class=" css-g1d714-ValueContainer"
        >
          <div
            class=" css-1wa3eu0-placeholder"
          >
            Enter the domains you want to send requests to here...
          </div>
          <div
            class="css-b8ldur-Input"
          >
            <div
              class=""
              style="display: inline-block;"
            >
              <input
                aria-autocomplete="list"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                id="react-select-2-input"
                spellcheck="false"
                style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
              <div
                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
              />
            </div>
          </div>
        </div>
        <div
          class=" css-1hb7zxy-IndicatorsContainer"
        />
      </div>
    </div>
    <div
      class="_1fo50kt _zjik7 _1h3rtzg"
    >
      <label
        class="_1gz641 _1efi8gv"
        for="warmupOnStart"
      >
        Post-registration warmup
      </label>
      <input
        id="warmupOnStart"
        type="checkbox"
      />
    </div>
  </div>
</DocumentFragment>
`;
